File: 23-merge-k-sorted-lists/23-merge-k-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        Queue<ListNode> q = new PriorityQueue<>(
            (ListNode n1, ListNode n2) -> Integer.compare(n1.val, n2.val));

        if(lists == null) return null;
        for(int i = 0; i < lists.length; i++) {
            if(lists[i] != null) q.offer(lists[i]);
        }

        ListNode prev = new ListNode();
        ListNode dummy = prev;
        while(!q.isEmpty()) {
            ListNode curr = q.poll();
            if (curr.next != null) q.offer(curr.next);
            prev.next = curr;
            prev = curr;
        }

        return dummy.next;
    }
}

File: 23-merge-k-sorted-lists/23-merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
from queue import PriorityQueue
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        class Wrapper():
            def __init__(self, node):
                self.node = node
            def __lt__(self, other):
                return self.node.val < other.node.val

        head = tail = ListNode(-1)
        pq = PriorityQueue()

        if not lists: return None
        for l in lists:
            if l:
                pq.put(Wrapper(l))

        while not pq.empty():
            node = pq.get().node
            tail.next = node
            # tail.next = ListNode(val)
            tail = tail.next

            if node := node.next:
                pq.put(Wrapper(node))

        return head.next